File: java/src/main/java/com/fish/string/Q3.java
package com.fish.string;

import java.util.HashSet;
import java.util.Set;

/**
 * @author liuqi
 * @date 2023/8/28
 * <p>
 * https://leetcode.cn/problems/longest-substring-without-repeating-characters/
 */
public class Q3 {
    class Solution {
        public int lengthOfLongestSubstring(String s) {
            int left = 0, right = 0;
            int maxLen = 0;
            Set<Character> set = new HashSet<>();
            while (right < s.length()) {
                if (!set.contains(s.charAt(right))) {
                    set.add(s.charAt(right));
                    maxLen = Math.max(maxLen, right - left + 1);
                } else {
                    while (s.charAt(left) != s.charAt(right)) {
                        set.remove(s.charAt(left));
                        left++;
                    }
                    left++;
                }
                right++;
            }
            return maxLen;
        }
    }
}

File: C++/Q3.cpp
#include <string>
#include <unordered_set>

using namespace std;

class Solution
{
public:
    int lengthOfLongestSubstring(string s)
    {
        if(s.size()==0){
            return 0;
        }
        int ans = 0;
        unordered_set<char> set;
        int left = 0;
        int right = 0;
        set.insert(s.at(left));
        while (left < s.size() && right < s.size())
        {
            while (right + 1 < s.size() && set.find(s.at(right + 1)) == set.end())
            {
                set.insert(s.at(right + 1));
                right++;
            }
            ans = max(ans, right - left + 1);
            set.erase(s.at(left));
            left++;
        }
        return ans;
    }
};